File: plugins/plugin-browser/typescript/src/utils/retry.ts
import { logger } from "@elizaos/core";
import type { RetryConfig } from "../types.js";

export const DEFAULT_RETRY_CONFIGS = {
  navigation: {
    maxAttempts: 3,
    initialDelayMs: 1000,
    maxDelayMs: 5000,
    backoffMultiplier: 2,
  },
  action: {
    maxAttempts: 2,
    initialDelayMs: 500,
    maxDelayMs: 2000,
    backoffMultiplier: 1.5,
  },
  extraction: {
    maxAttempts: 2,
    initialDelayMs: 500,
    maxDelayMs: 3000,
    backoffMultiplier: 2,
  },
} satisfies Record<string, RetryConfig>;

export async function retryWithBackoff<T>(
  fn: () => Promise<T>,
  config: Partial<RetryConfig> & { timeout?: number },
  operation: string
): Promise<T> {
  const maxAttempts = config.maxAttempts ?? 3;
  const initialDelay = config.initialDelayMs ?? 1000;
  const maxDelay = config.maxDelayMs ?? 5000;
  const backoffMultiplier = config.backoffMultiplier ?? 2;

  let lastError: Error | undefined;
  let delay = initialDelay;

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      logger.info(`Attempting ${operation} (attempt ${attempt}/${maxAttempts})`);

      if (config.timeout) {
        const timeoutPromise = new Promise<never>((_, reject) =>
          setTimeout(
            () => reject(new Error(`${operation} timed out after ${config.timeout}ms`)),
            config.timeout
          )
        );
        return await Promise.race([fn(), timeoutPromise]);
      }

      return await fn();
    } catch (error) {
      lastError = error as Error;
      const errorMessage = error instanceof Error ? error.message : String(error);
      logger.warn(`${operation} failed (attempt ${attempt}/${maxAttempts}): ${errorMessage}`);

      if (attempt < maxAttempts) {
        logger.info(`Retrying ${operation} in ${delay}ms...`);
        await new Promise((resolve) => setTimeout(resolve, delay));
        delay = Math.min(delay * backoffMultiplier, maxDelay);
      }
    }
  }

  logger.error(`${operation} failed after ${maxAttempts} attempts`);
  throw lastError ?? new Error(`${operation} failed after ${maxAttempts} attempts`);
}

export function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

File: plugins/plugin-browser/rust/src/utils/retry.rs
use crate::types::RetryConfig;
use std::future::Future;
use std::time::Duration;
use tokio::time::sleep;
use tracing::{info, warn};

pub mod default_configs {
    use super::RetryConfig;

    pub fn navigation() -> RetryConfig {
        RetryConfig {
            max_attempts: 3,
            initial_delay_ms: 1000,
            max_delay_ms: 5000,
            backoff_multiplier: 2.0,
        }
    }

    pub fn action() -> RetryConfig {
        RetryConfig {
            max_attempts: 2,
            initial_delay_ms: 500,
            max_delay_ms: 2000,
            backoff_multiplier: 1.5,
        }
    }

    pub fn extraction() -> RetryConfig {
        RetryConfig {
            max_attempts: 2,
            initial_delay_ms: 500,
            max_delay_ms: 3000,
            backoff_multiplier: 2.0,
        }
    }
}

pub async fn retry_with_backoff<T, E, F, Fut>(
    mut f: F,
    config: &RetryConfig,
    operation: &str,
) -> Result<T, E>
where
    F: FnMut() -> Fut,
    Fut: Future<Output = Result<T, E>>,
    E: std::fmt::Debug,
{
    let mut delay = config.initial_delay_ms;
    let mut last_error: Option<E> = None;

    for attempt in 1..=config.max_attempts {
        info!(
            "Attempting {} (attempt {}/{})",
            operation, attempt, config.max_attempts
        );

        match f().await {
            Ok(result) => return Ok(result),
            Err(e) => {
                warn!(
                    "{} failed (attempt {}/{}): {:?}",
                    operation, attempt, config.max_attempts, e
                );
                last_error = Some(e);

                if attempt < config.max_attempts {
                    info!("Retrying {} in {}ms...", operation, delay);
                    sleep(Duration::from_millis(delay)).await;
                    delay = std::cmp::min(
                        (delay as f64 * config.backoff_multiplier) as u64,
                        config.max_delay_ms,
                    );
                }
            }
        }
    }

    tracing::error!(
        "{} failed after {} attempts",
        operation,
        config.max_attempts
    );
    Err(last_error.unwrap())
}

pub async fn sleep_ms(ms: u64) {
    sleep(Duration::from_millis(ms)).await;
}
